tools/test/CoverageOutputGenerator/java/com/google/devtools/coverageoutputgenerator/GcovJsonParser.java

```diff
@@ -0,0 +1,123 @@
+  private final InputStream inputStream;
+
+  private GcovJsonParser(InputStream inputStream) {
+    this.inputStream = inputStream;
+  }
+
+  public static List<SourceFileCoverage> parse(InputStream inputStream) throws IOException {
+    return new GcovJsonParser(inputStream).parse();
+  }
+
+  private List<SourceFileCoverage> parse() throws IOException {
+    ArrayList<SourceFileCoverage> allSourceFiles = new ArrayList<>();
+    try (InputStream gzipStream = new GZIPInputStream(inputStream)) {
+      ByteArrayOutputStream contents = new ByteArrayOutputStream();
+      byte[] buffer = new byte[1024];
+      int length;
+      while ((length = gzipStream.read(buffer)) != -1) {
+        contents.write(buffer, 0, length);
+      }
+      Gson gson = new Gson();
+      GcovJsonFormat document = gson.fromJson(contents.toString(), GcovJsonFormat.class);
+      if (!document.format_version.equals("1")) {
+        logger.log(Level.WARNING, "Expect GCov JSON format version 1, got format version " + document.format_version);
+      }
+      for (GcovJsonFile file : document.files) {
+        SourceFileCoverage currentFileCoverage = new SourceFileCoverage(file.file);
+        for (GcovJsonFunction function : file.functions) {
+          currentFileCoverage.addLineNumber(function.name, function.start_line);
+          currentFileCoverage.addFunctionExecution(function.name, function.execution_count);
+        }
+        for (GcovJsonLine line : file.lines) {
+          currentFileCoverage.addLine(line.line_number, LineCoverage.create(line.line_number, line.count, null));
+          for (GcovJsonBranch branch : line.branches) {
+            currentFileCoverage.addBranch(line.line_number, BranchCoverage.create(line.line_number, branch.count));
+          }
+        }
+        allSourceFiles.add(currentFileCoverage);
+      }
+    }
+
+    return allSourceFiles;
+  }
+
+  // Classes for the Gson data mapper representing the structure of the GCov JSON format
+  // These do not follow the Java naming styleguide as they need to match the JSON field names
+  // Documentation can be found in GCov's manpage, of which the source is available at
+  // https://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/doc/gcov.texi;h=dcdd7831ff063483d43e5347af0b67083c85ecc4;hb=4212a6a3e44f870412d9025eeb323fd4f50a61da#l184
+
+  class GcovJsonFormat {
+    String gcc_version;
+    GcovJsonFile[] files;
+    String format_version;
+    String current_working_directory;
+    String data_file;
+  }
+
+  class GcovJsonFile {
+    String file;
+    GcovJsonFunction[] functions;
+    GcovJsonLine[] lines;
+  }
+
+  class GcovJsonFunction {
+    int blocks;
+    int end_column;
+    int start_line;
+    String name;
+    int blocks_executed;
+    int execution_count;
+    String demangled_name;
+    int start_column;
+    int end_line;
+  }
+
+  class GcovJsonLine {
+    GcovJsonBranch[] branches;
+    int count;
```

Comment:
The type of count could be long?   
see : https://github.com/bazelbuild/bazel/issues/9406#issuecomment-664211209